libraries/lib-basic-ui/BasicUI.h

```diff
@@ -0,0 +1,289 @@
+   ProgressShowStop            = (1 << 0),
+   ProgressShowCancel          = (1 << 1),
+   ProgressHideTime            = (1 << 2),
+   ProgressConfirmStopOrCancel = (1 << 3),
+};
+
+enum class ProgressResult : unsigned
+{
+   Cancelled = 0, //<! User says that whatever is happening is undesirable and shouldn't have happened at all
+   Success,       //<! User says nothing, everything works fine, continue doing whatever we're doing
+   Failed,        //<! Something has gone wrong, we should stop and cancel everything we did
+   Stopped        //<! Nothing is wrong, but user says we should stop now and leave things as they are now
+};
+
+//! Abstraction of a progress dialog with well defined time-to-completion estimate
+class BASIC_UI_API ProgressDialog
+{
+public:
+   virtual ~ProgressDialog();
+
+   // Update the bar and poll for clicks
+   virtual ProgressResult Poll(
+      unsigned long long numerator,
+      unsigned long long denominator,
+      const TranslatableString &message = {}) = 0;
+};
+
+//! Abstraction of a progress dialog with undefined time-to-completion estimate
+class BASIC_UI_API GenericProgressDialog
+{
+public:
+   virtual ~GenericProgressDialog();
+   //! Give some visual indication of progress
+   virtual void Pulse() = 0;
+};
+
+//! @}
+
+//! Abstract class defines a few user interface services, not mentioning particular toolkits
+/*! The intention is that the application supplies a concrete implementation at
+ startup.  Most code will not use this class directly, but call the inline
+ functions that follow. */
+class BASIC_UI_API Services {
+public:
+   virtual ~Services();
+   virtual void DoCallAfter(const Action &action) = 0;
+   virtual void DoYield() = 0;
+   virtual void DoShowErrorDialog(const WindowPlacement &placement,
+      const TranslatableString &dlogTitle,
+      const TranslatableString &message,
+      const ManualPageID &helpPage,
+      const ErrorDialogOptions &options) = 0;
+   virtual MessageBoxResult DoMessageBox(
+      const TranslatableString& message,
+      MessageBoxOptions options) = 0;
+   virtual std::unique_ptr<ProgressDialog>
+   DoMakeProgress(const TranslatableString &title,
+      const TranslatableString &message,
+      unsigned flag,
+      const TranslatableString &remainingLabelText) = 0;
+   virtual std::unique_ptr<GenericProgressDialog>
+   DoMakeGenericProgress(const WindowPlacement &placement,
+      const TranslatableString &title,
+      const TranslatableString &message) = 0;
+};
+
+//! Fetch the global instance, or nullptr if none is yet installed
+BASIC_UI_API Services *Get();
+
+//! Install an implementation; return the previously installed instance
+BASIC_UI_API Services *Install(Services *pInstance);
+
+//! @name Functions that dispatch to the global Services, if supplied, or else are no-ops
+//! @{
+
+//! Schedule an action to be done later in the event loop, and in the main thread
+inline void CallAfter(const Action &action)
+{
+   if (auto p = Get())
```

Comment:
Why not throw an exception here? Or assert that BasicUI::Get() isn't null? Considering client code scheduling some job while we don't have a UI instance.

Reply:
Assertion would be wrong -- there is no proof that services must be installed before calling CallAfter.  That is not stated as a precondition anywhere.

Exception would be better, then.  But then that must be part of the stated contract of this function, and then anything using CallAfter would require a catch block, in it or at some higher level.

But an alternative might be that the no-op implementation saves all Actions in a list, and calls them (only) when Yield is called, and that could be stated in the comments as default behavior.  Then default CallAfter behavior could at least not lose the actions (but, require you to do something else to guarantee that they happen).

In practice this won't matter, given that we do install services in Audacity, but you are right to think about this library in isolation from that assumption about its usage.